### DISCUSSION TITLE
Integration with GraalVM native images

### DISCUSSION THREAD

**Discussion by ennerf** [hydraulic-software/conveyor#66] [Ideas]
I'm currently experimenting with migrating a JRE+multi-launcher Conveyor package to use GraalVM native image. I am able to create a native shared library for multiple launchers (see [native-launchers](https://github.com/HebiRobotics/native-launchers-maven-plugin)), but using a [custom launcher template](https://github.com/HebiRobotics/native-launchers-maven-plugin/blob/2394fdb9c724d54c78a07b3177b049e266be9750/native-launchers-maven-plugin/src/main/resources/us/hebi/launchers/templates/main-dynamic.c) removes some of the benefits that Conveyor provides, e.g.,

1) Missing properties like `-Dapp.dir` to figure out where to load resources from

2) The launchers do not check for updates

3) It's hard to implement manual update checks due to required JNI calls/libs to Sparkle

Have you considered how an integration with native-images could look like, and/or is there a way to solve these issues from the user-side?

**Comment by mikehearn** [comment]
Yes, we've looked at it. Conveyor 12 onwards has some fixes that let us treat a native image as if it's a special kind of JVM app. There's no properly integrated support in the product yet, but interest is growing as native image matures.

Here's some info on the current state.

Firstly you have to compile your app with `--shared` and some JSON configs to ensure JNI and reflection work. You also have to add a JAR from the Conveyor distribution with some startup code. Use an incantation of this rough form:

```java
class Hello {
	public static void main(String[] args) {
		System.out.println("hello world");
	}
}
```

```shell
$ javac Hello.java

$ native-image -H:JNIConfigurationFiles=jni.json -H:ReflectionConfigurationFiles=reflect.json --shared -cp .:/Applications/Conveyor.app/Contents/Resources/launchers/launcher.jar Hello hello
```

`jni.json`:

```json 
[
	{
		"name": "hydraulic.conveyor.launcher.Launcher",
		"allDeclaredMethods": true,
		"allDeclaredFields": true
	},

	{
		"name": "Hello",
		"allDeclaredMethods": true
	}

]
```

`reflect.json`:

```json
[
	{
		"name": "Hello",
		"allDeclaredMethods": true
	}
]
```

It's a bit indirect because the Conveyor C++ side uses Java to look up and possibly instantiate the main class (to handle JavaFX apps that inherit from `Application` and to support other features in future), then JNI to invoke the looked up main method. Native image has to be told about this to work. A possible improvement is to just skip this JNI stuff for native images and invoke the main class function directly. However, then we'd lose the ability to run Java code before the user's app.

Secondly, it requires some config hacks which a product-quality version of this feature would need to resolve:

- You have to create a little fake JDK where your native image is placed in the place `libjvm` would be found. This fake JDK needs a minimal `release` file too. Fortunately Conveyor's input handling is powerful enough that you don't have to actually create this on disk: a few lines of config is sufficient.
- You need a fake input to get past the "you forgot to add inputs" safety check. We'd have to teach Conveyor that it's OK for an app to have no base inputs when they're all rolled into the JVM binary itself. Unfortunately the "no inputs" check does happen earlier than any JVM specific stuff so that needs a bit of thought.
- You need to bump your minimum macOS version higher than the default, because native-image doesn't set any min so it defaults to whatever OS you're running on. This is probably fixable with some native-image linker flags.

```hocon
app {
	jvm {
		gui.main-class = Hello

		mac.amd64.inputs = [
			hello.dylib -> Contents/Home/lib/server/libjvm.dylib

			{
				to = Contents/Home/release
				content = """
					JAVA_VERSION="20.0.1"
					MODULES=""
					OS_ARCH="amd64"
					OS_NAME="mac"
				"""
			}
		]
	}

	fsname = native-image-test
	version = 1.0
	site.base-url = "localhost:80"
	rdns-name = com.example.native-image

	inputs = [
		{
			to = foo
			content = .
		}
	]

	// Native image produces binaries that use the latest mac SDK by default.
	mac {
		info-plist.LSMinimumSystemVersion = 13.0.0
	}
}
```

With this I can get a Mac bundle for a native image that builds and launches in the regular way and prints Hello World. There's some roughness along the way - it'll warn me there are no JARs, and it downloads a JLink JDK even though it doesn't actually use it for anything. But it does seem to work.

Extending it to Windows and Linux is presently left as an exercise for the reader, but anyone who has commercial demand for this should let us know.

**Comment by ennerf** [comment]
Now that it's a few years later, I'm finally at a point where I can package all of my tools into a single shared native-image with a few tiny wrapper launchers.

With current version it's easy enough to package it as a native-app, but is there a way to get to get access to the variables that Conveyor would normally set? e.g. the `app.windows.userModelID` (AUMID?) and maybe the various cross-platform app directories?

I'm also used to having a button to trigger an update via the conveyor API. Can I do something similar in a native app without just running `updatecheck.exe` ?

**Reply by mikehearn** [reply]
Those are system properties and should be set automatically, as we're passing them in to the JVM startup code via the Conveyor generated launchers. Do they not show up for you?

You should just be able to use the regular Conveyor update control API, I think? There's no need for any JNI config there, as it just calls out to native code.

**Reply by ennerf** [reply]
I'd like to switch to native image partly to avoid the overhead of starting a vm, so I'm creating custom [native launchers ](https://github.com/HebiRobotics/native-launchers-maven-plugin) wrapping the native-image. I can work around most of the missing `app.*` variables by embedded some custom properties and the path to the current executable, but it'd obviously be nicer to just have Conveyor generated launchers w/ icons and properties.

I haven't quite gotten to the point of full integration, but I'll keep you updated in case it doesn't work.

Native shared-libraries actually provide the same JNI API as a normal VM (see [launcher_dynamic.c](https://github.com/HebiRobotics/native-launchers-maven-plugin/blob/main/native-launchers-maven-plugin/src/main/resources/us/hebi/launchers/templates/launcher_dynamic.c#L148-L163)). If you wanted to offer first party support and run custom Java-code first, I think you could do it by providing a runtime lib w/ bundled reflect-config that gets compiled into the native-image.

Users would just need to supply a graalvm lib input combined with a main-entry-point or a class name that is registered for reflection.

**Reply by mikehearn** [reply]
Yes, I think the instructions above use that capability, no? That is, you'd compile your apps into one single shared library using native image, and then Conveyor creates its own launchers that use the JNI API to start up the "virtual machine" i.e. app with the correct main class, system properties, etc. That's the solution I described above and I think it worked at the time.

I've been wanting to revisit native image support for a while. Hopefully that will be possible soon, as most higher priority items have been done.

**Reply by ennerf** [reply]
Well, yes. After re-reading that's pretty much what I was thinking 🤦

I just tried it on Windows, but I'm not sure what files it needs to treat my fake jvm as a full jdk. I copied the release file of the used jdk, but that didn't seem to be sufficient. 

```
❌ Task failed: Windows App for Intel: JAR files were imported but no JVM inputs were supplied for 'windows.amd64'.

   Did you import a JDK? Try adding `include "/stdlib/jdk/17/amazon.conf"` to your
   config file if not.
```

It looks like missing jar inputs also changed from a warning to an error? Without a jar I get

```
❌ Task failed: Windows App for Intel: No EXE files were found in the root of the Windows inputs.
```

Anyways, if I build it with a full jdk and replace the `jvm.dll` with my renamed Graal binary (w/ launcher.jar and jni/reflect-config), I get a `Failed to create JVM: Error 0xffffffff` error. Not particularly descriptive, but I assume that fails on the initial creation before it even gets to the reflective calls? Is there any way I can help debug that?

**Reply by mikehearn** [reply]
That is the return code from JNI_CreateJVM which translates to the rather unhelpfu JNI_ERR a.k.a the notorious "unknown error".

After a bit of digging we find the remarkable `-H:+JNIEnhancedErrorCodes` option to native-image, which is documented here:

https://github.com/oracle/graal/blob/b8c76e4cf88c240ab22808a4f0e242c7b08f49d7/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java#L712

"JNI functions will return more specific error codes"

Why this has to be turned on explicitly is mysterious to me... but it might help make progress. SubstrateVM has a special protocol for communicating string errors to the caller of JNI_CreateJVM. Implementing support for that would be a good start.

To hack around the other problems you can just add an empty jar file for now. How are you specifying your config file? The error means there was no JDK specified, which is expected, but if you create a fake JDK with a release file that test should pass.

**Reply by mikehearn** [reply]
Actually I've reproduced that error (on macOS at least).  It happens when a native image isn't identified as such. You should see a warning saying that native image support is experimental during the build, but I bet you don't.

I'm working on better native image support at the moment and will fix some issues with this.

**Reply by ennerf** [reply]
> I'm working on better native image support at the moment and will fix some issues with this.

Awesome! Let me know if I can help beta test something.

> To hack around the other problems you can just add an empty jar file for now. How are you specifying your config file? The error means there was no JDK specified, which is expected, but if you create a fake JDK with a release file that test should pass.

The config looked something like this:

```hocson
app {

  windows.amd64.package-extras = [
      target/image/hebiscope.dll -> bin/jvm.dll // pretend it's a Java runtime
      // ${env.JAVA_HOME}/release -> release // make Conveyor think it's a real jdk
      {
        to = release
        content = """
            JAVA_VERSION="21.0.8+13-LTS"
            MODULES=""
            OS_ARCH="x86_64"
            OS_NAME="Windows"
         """
      }
    ]

  inputs += target/gui-0.1-SNAPSHOT.jar // dummy jar to keep it from complaining

}
```

**Reply by ennerf** [reply]
The [macOS update trigger](https://conveyor.hydraulic.dev/21.0/control-api-native/#macos) turns out to be a bit of a pain to get working. The Conveyor jar didn't work out of the box, JNI was a pain, JNA didn't work, and FFM wasn't supported in my native-image version.

edited: it looks like the GraalVM C interface does not need the symbol to be available during image build time, so this works:

```Java
public class ConveyorMacOS {
    @CFunction(value = "conveyor_check_for_updates", transition = CFunction.Transition.NO_TRANSITION)
    public static native void invokeUpdateCheck();
}
```

edit2: well, almost... all GUI apps work, but CLI apps don't seem to add extra library

```text
dlopen(hebiscope.dylib, 0x0001): symbol not found in flat namespace '_conveyor_check_for_updates'
```

edit3: Here is an initial version calling dlopen/dlsym via the native-image C API. I'm sure someone more familiar with native code can figure out a significantly simpler solution, but for now this does what I need.

```Java
public class ConveyorMacOS {

    public static void checkForUpdates() {
        if (!Platform.includedIn(Platform.MACOS.class)) {
            throw new UnsupportedOperationException("Conveyor update check is only supported on macOS.");
        }

        try (CTypeConversion.CCharPointerHolder symbolName = CTypeConversion.toCString("conveyor_check_for_updates")) {

            // Passing NULL to dlopen searches the current process/loaded libs
            // (libconveyor is already injected into the executable header)
            final int RTLD_LAZY = 1;
            PointerBase handle = LibDL.dlopen(WordFactory.nullPointer(), RTLD_LAZY);
            if (handle.isNull()) throw new AssertionError("dlopen failed");

            // Look up the symbol
            PointerBase funcPtr = LibDL.dlsym(handle, symbolName.get());
            if (funcPtr.isNull()) {
                CCharPointer error = LibDL.dlerror();
                var msg = error.isNull() ? "" : CTypeConversion.toJavaString(error);
                throw new RuntimeException("failed to find symbol: " + msg);
            }

            // Call native function
            ((CheckForUpdatesFunctionPointer) funcPtr).invoke();

        } catch (Exception e) {
            throw new RuntimeException("failed to invoke update check", e);
        }

    }

    interface CheckForUpdatesFunctionPointer extends CFunctionPointer {
        @InvokeCFunctionPointer(transition = CFunction.Transition.NO_TRANSITION)
        void invoke();
    }

    @Platforms(Platform.MACOS.class)
    static class LibDL {
        // Define POSIX dlopen (only available on macOS/Linux)
        @CFunction(transition = CFunction.Transition.NO_TRANSITION)
        public static native PointerBase dlopen(CCharPointer filename, int flags);

        @CFunction(transition = CFunction.Transition.NO_TRANSITION)
        public static native PointerBase dlsym(PointerBase handle, CCharPointer symbol);

        @CFunction(transition = CFunction.Transition.NO_TRANSITION)
        private static native CCharPointer dlerror();
    }

}
```

**Reply by mikehearn** [reply]
It's interesting you needed that, the library should be usable via JNI out of the box with native-image. Do you recall what the JNI issues were when you tried it?

**Reply by ennerf** [reply]
I unfortunately don't remember the details and can't find my notes on it. I do remember that the bundled control library didn't work out of the box, and creating custom JNI bindings would have added yet another native build/compilation step to an already convoluted build setup. If I remember correctly I also ran into some issues with not having the library/symbol at native-image compile time.

I'm also generally very unfamiliar with JNI, so you might be able to figure out something simpler 🤷
